### ISSUE TITLE
Analysis of the bad moves in the Golaxy series G6

### ISSUE THREAD

**Issue by bjiyxo** [gcp/leela-zero#1226]
Please don't row here. I want to discuss technical issue.
 
Definitely there is a bug on [V9](http://zero.sjeng.org/networks/32179366aa6220fa85ff98f8813b7d7663a0339ebfaf308dad9c3693a22a0d22.gz) (32179366) in [G6](http://portal.yikeweiqi.com/online/golive/detail?id=8405). The losing move is move 19, and V9 should play C6 or P16. In the real game, V9 plays C7, and the upper right corner dies suddenly.
Here is the log from V9 with 300s thinking time.
```
  C7 ->   99917 (V: 61.66%) (N: 84.43%) PV: C7 M4 P16 R11 D7 E7 E8 F8 F7 E6 E9 G7 F9 F6 Q4 Q3 O4 Q5 H8 M6 H7 P5 K3 L3 H5 H3 K4 K7 K9
  C6 ->     348 (V: 58.35%) (N:  4.82%) PV: C6 D7 P16 Q11 C8 L4 D8 E8 E9 F9 F8 E7 E10 F10 F11 E11 C7
 P16 ->     167 (V: 54.41%) (N:  4.94%) PV: P16 C6 R11 Q12 Q11 P11 P10 R12 S12 O11 O10 N11
  B7 ->      12 (V: 59.09%) (N:  0.11%) PV: B7 P18 P16 O18 N18 M18 N17
```

Later I found out that [V4](https://drive.google.com/file/d/1y1A-ijeE1f50EuT0haeWUEtNdtUhJuON/view?usp=sharing) and [V4swa](https://drive.google.com/file/d/11GizcSzrr3ZTx8GPOOdlSn8b5YdWdN4R/view?usp=sharing) can easily find the exact move, while [V5](https://drive.google.com/file/d/1TikFfZkrtzfs0jBu1Y7-4IE-T6TMXRRY/view?usp=sharing), [V6](https://drive.google.com/file/d/1sWfBzeXJsuK6xCGnCrpfNTx6R1mHs66F/view?usp=sharing), [V7](https://drive.google.com/file/d/1fx4DeqLDUuC2BlyWqY4ubgFtUDL6GNi9/view?usp=sharing), and [V8](https://drive.google.com/file/d/1HpEQwzhDlgp0v636aSkY7ha_hDbtX7u0/view?usp=sharing)  can find out with much more time and V9 cannot.
Here are the logs from V4, V4swa, V5, V6, V7, V8 with 300s thinking time.
V4
```
   P16 ->   96440 (V: 53.15%) (N:  2.32%) PV: P16 C6 R12 Q13 R9 Q12 R11 R6 P10 M14 Q5 Q6 Q3 R4 Q4 S7 M3 N11 C2 O9 P9
  C6 ->   31728 (V: 52.48%) (N:  4.24%) PV: C6 D7 P16 C7 D2 R11 E2 Q5 Q10 Q11 P10 C12 R10 P11 O10 D16 E16 C16 C17 B17 F17 B15 O11 S11 O13
  D3 ->    5516 (V: 52.49%) (N:  0.57%) PV: D3 D2 E4 D4 C3 F3 C7 D7 C8 D8 D9 E9 E10 F9 F10 G9 G10 C6 B6 B5 B7 C2 B4 H9 M3 O3 J3 Q9 H10
  C7 ->    1593 (V: 37.21%) (N: 91.51%) PV: C7 P18 P16 O18 M17 N18 L16 S15 Q18 S18 M18 S17 D7 J3 N17 R18 F7 F17 G3 J17
  B7 ->      21 (V: 51.06%) (N:  0.08%) PV: B7 P18 P16 O18 N18 N17 N16 M17 M18 L18 L17 M16 K18
```

V4swa
```
  C6 ->   99933 (V: 59.43%) (N:  2.71%) PV: C6 D7 P16 R11 C8 L4 Q4 Q3 O4 Q5 P5 R5 P7 M6 Q10 Q8 O9 Q11 P11 P12 O11 O12 D8 E8 E9 F9 F8 E7 E10
  C7 ->    1851 (V: 46.87%) (N: 91.03%) PV: C7 P18 P16 O18 N18 N17 M17 Q18 S15 S14 S16 R18 R14 R13 Q14 P15 N16 Q13 S18 S19 T18 T14 T15 T17
 P16 ->     144 (V: 51.47%) (N:  4.50%) PV: P16 C6 R11 Q12 Q11 P11 P10 R12 O10 O11 N11 N10
  D3 ->      13 (V: 56.71%) (N:  0.14%) PV: D3 D2 C2 D4 C3 C6 E2 F2
```

V5
```
  C6 ->   99955 (V: 60.05%) (N:  4.39%) PV: C6 D7 P16 Q11 C8 L4 Q6 Q8 Q4 Q3 O4 R4 D8 P5 F8 G7 O6 M6 P6 O8 N7 L8
  C7 ->   28325 (V: 58.86%) (N: 90.72%) PV: C7 Q9 D7 E7 E8 F8 F7 E6 E9 G7 F9 F6 P16 M4 H8 Q11 P14 P13 P15 Q13 Q4 R4 Q6 P4 Q8 P8
 P16 ->      92 (V: 50.87%) (N:  3.12%) PV: P16 C6 R11 Q12 Q11 P11 P10 R12 S12 O11 O10
  D3 ->       6 (V: 55.98%) (N:  0.10%) PV: D3 D2 C2 D4
```

V6
```
  C6 ->   99966 (V: 58.32%) (N:  2.84%) PV: C6 D7 P16 Q11 C8 L4 Q6 Q8 P4 Q4 P5 P3 O3 P2 M5 L5 M6 L6 M7 L7 M3 L3 D8 M8 N8 N9 O8 O9 P8 P9
  C7 ->   32722 (V: 57.15%) (N: 93.96%) PV: C7 L4 P16 Q11 D7 E7 E8 F8 F7 E6 E9 F9 D4 E4 F6 F10 E10 F11 E11 H4 F12 G12 G13 H12 H13 J12 J13
 P16 ->      84 (V: 51.17%) (N:  2.18%) PV: P16 C6 R11 Q12 Q11 P11 P10 R12 S12 O11 O10
```

V7
``` 
  C6 ->   99927 (V: 56.96%) (N:  4.53%) PV: C6 D7 P16 Q11 C8 L4 Q4 Q3 O4 Q5 D8 E8 E9 F9 F8 E7 E10 F10 E11 G8 M4 L5 L3 K3 K2 L2 M3 J2 K4 J3 M6 C15 E16 C17 C18 B18 B17 C16 B19 A18 E18 C13
  C7 ->   67877 (V: 56.54%) (N: 88.74%) PV: C7 L4 P16 R11 D7 E7 E8 F8 F7 E6 E9 F9 D4 E4 F6 F10 E10 F11 E11 H4 F12 G12 G13 H12 H13 J12 J13 K12 K13 L13 L14 M14 L15 C17 C16 D16 E17
 P16 ->     157 (V: 51.66%) (N:  2.91%) PV: P16 C6 R11 Q12 Q11 P11 P10 R12 S12 S13 S11
```

V8
``` 
  C6 ->   99866 (V: 58.01%) (N:  4.97%) PV: C6 D7 C8 L4 Q4 Q3 O4 Q5 Q11 Q9 O11 O9 P16 Q13 N13 N14 M14 M15 O13 O15
  C7 ->   68415 (V: 57.53%) (N: 84.97%) PV: C7 L4 P16 D7 C8 R11 D8 Q6 F8 G7 N3 M3 M5 L5 M6 L6 M7 N2 L7 N4 J7 G8
 P16 ->     279 (V: 53.08%) (N:  4.82%) PV: P16 C6 R11 Q12 Q11 P11 P10 R12 S12 O11 O10 S13 S10
```
It looks like the weights forget the exact sequence little by little.

And the followings are the things I do between V4swa and V9.
1. I didn't turn off swa so the amount of swa still counts 16 until the end of V9.
2. I trained ~550k steps, and the window shifted ~200k games between each version.
3. The batch size is 128 before V6 and 512 after V6. The learning rate is 0.0001.

I'm not sure what I did wrong. Any suggestion?

**Comment by isty2e** [comment]
Current 15b network does not play C7, so I think it is not related much to the training set, though it can slightly differ. How large is the window?

**Comment by bood** [comment]
IIRC v9 is the first network trained by multiple GPU code, maybe there is a bug somewhere?

**Comment by herazul** [comment]
Maybe restarting from a net2net of last 15b and training from 15b selfplay games would be a solution, but that would set you back on level for a week.

**Comment by wentaol** [comment]
Is there any difference in training methods between these weights and the official 15b weights?

**Comment by l1t1** [comment]
sgf
http://duiyi.sina.com.cn/cgibo/20184/leela6-golaxy180419.sgf

**Comment by Splee99** [comment]
I don't know why O16 was played (move 13) during the game, which I guess was the root cause of the problem. This is what the latest weight prefers:
```text
Playouts: 6255, Win: 57.09%, PV: O17 D5 C5 D6 C7 Q9 D7 O4 F7 H3 Q5 P4 R4 S3 S9 R8 S8
 O17 ->    6266 (V: 57.14%) (N: 89.18%) PV: O17 D5 C5 D6 C7 Q9 D7 O4 F7 H3 Q5 P4 R4 S3 S9 R8 S8
 O16 ->      92 (V: 56.09%) (N:  2.81%) PV: O16 D5 C5 D6 C7 O17 P17 N17 P16 K17 R13 Q13 Q12
 N17 ->      26 (V: 56.48%) (N:  0.65%) PV: N17 D5 C5 D6 C7 Q9 D7 L3 F7 G6
  D6 ->       8 (V: 50.18%) (N:  0.96%) PV: D6 O17 O18 N18 P18 M17 H3 S15
12.4 average depth, 32 max depth
4745 non leaf nodes, 1.35 average children
6393 visits, 2164299 nodes, 6392 playouts, 108 n/s

**Comment by bubblesld** [comment]
I remember that direct training from the latest game usually fails. I guess that LZ needs the early games to learn from some "wrong" moves. When the weights are updated using the current self-play games, it tries to address the weakness from self-play. Since those "trivial" mistakes will not happen from the current way, the feature captured earlier will gradually be ignored.

I have two suggests to avoid what I mentioned above:
1. Maybe we should randomly include some early games (eq. 10%) when training with new games.
2. For self-play games, currently there are some randomness for the first 30 moves. How about the randomness comes from early version? I mean, choosing a previous version at random, then for the first 30 moves, there is some chance (say, 10%) that the first 30 moves are generated by the previous version.

**Comment by bjiyxo** [comment]
> IIRC v9 is the first network trained by multiple GPU code, maybe there is a bug somewhere?
No. It's from v7.

**Comment by bjiyxo** [comment]
> I don't know why O16 was played (move 13) during the game, which I guess was the root cause of the problem

Yes. Almost all weights don't have the problem, including 6b. So I think there might be some problems in my training here.

**Comment by gcp** [comment]
Bad moves aren't bugs. They might be caused by them, but as long as we are not at perfect play there's no necessary correlation.

**Comment by herazul** [comment]
I was wondering something : 20x256 is training on 15b games but it doesnt give the same results, and don't play the same obviously. So it might be very well possible that it learn "mistakes" : there is no reason it wouldnt from time to time, the 15b does too. But what is different from the 15b is that it doesnt do self-play, so it can't play the mistake and learn from it.

Whereas a new 15b net better than the previous one but containing a mistake in a sequence might do it from time to time on self-play and learn from it.

**Comment by l1t1** [comment]
any analysis for game 7th?

**Comment by kuba97531** [comment]
I did analysis with reltively high playouts and LZ125 network.

http://portal.yikeweiqi.com/online/fight/online?id=3559342

Golaxy starts as black in tengen. The first move of Leela is weird (6th line).

Then the game progresses more or less as would Leela self-play with leela believing it's way ahead and suddenly it missed move 41 of golaxy (S8) and the winrate dropped below 50%.
Fast forward few obvious moves and Golaaxy is already at 60% after move 55. Fast forward few more moves and white is already at 31%. Leela plays tenuki from a life and death situation (my version agrees), but after black captures the group in the corner the game is at 85% for golaxy.

In short - except for the very first move Leela played according to what my LZ125 was suggesting (or very close second candidates). Golaxy was playing Leela moves 90% of the time, but when it diverged then LZ winrate dived very fast and the game was over after the first fight.

For entertainment here is a video in Lizzie showing how Leela after suprising opponent's move slowly realizes that it's position gets worse as more playouts come in.
https://gyazo.com/c931433d9cdf19d9b663c2b37398534b

**Comment by hydrogenpi** [comment]
@kuba97531 I concur that I too when attempting to duplicate or replicate the results, I could not get repeat nor duplicate LZ's first move no matter which network I tried....

So

Here is my theory, since we know the following facts :

1) The LZ moves are not automatically inputted into the game/match against its opponent Golaxy
2) It is a manual process that requires a human hand to 'click' and input the moves
3) There were on several occasions "operator misclicks" from on the volunteer LZ team side
4) There was announced via translation that game 1 had an 'error of the connector' which indicated to mean possibly another human input error. 

One can permissible infer and/or reasonably speculate from the above, given the fact that it seems multiple people cannot duplicate the first move from LZ's side in the 7th game, that it seems likely that either LZ volunteer team again 'misclicked' and/or they deliberately may have chosen to take a sort of creative liberties, if you will, in order to then start the game off an unusual path....

**Comment by fredfredc** [comment]
@kuba97531  @hydrogenpi 

I agree that the first move of LZ is  weird in game 7. It seems that LZ's first move was intervened by operator probably.

**Comment by l1t1** [comment]
game7 http://duiyi.sina.com.cn/cgibo/20184/leela7-golaxy180420.sgf

**Comment by pheasant75** [comment]
Golaxy learns from games played by human.  So first move tengen is no surprise, nor is it surprise that it played several stone above 6th line in one game.  That’s one style that some human players play.
G7 first move by LZ is the operator’s personal choice.  Since golaxy took a disadvantage, he doesn’t want to accept that, so he also plays above 6th line.  I believe subsequent move are LZ’s move.
These games do revealed some issues of LZ.  Seems it still have some blind spots.  Hope future training can improve thing.

**Comment by tterava** [comment]
It seems to me that LZ makes certain (correct) assumptions about joseki when it plays a game from start to finish. I've never seen 192x15 network do a simple ladder mistake in an even game where it has full control from the start. However if there's handicap stones or you manually place the first stones to mix up the fuseki, it's very easy for it to fall into ladder.

Basically what I'm saying is that forcefully mixing up LZ's fuseki probably makes it a lot weaker and I'm not sure if such mistakes should be considered bugs or flaws in the design. Doesn't it make sense that the network would assume some things in the early game to spare more capacity for later when it's more important?

**Comment by fredfredc** [comment]
Obviously , LZ's first move was played by human operator in game 7.

**Comment by hydrogenpi** [comment]
Next friday against Ke Jie

http://english.cctv.com/2018/04/20/ARTICxnU1RIDPklaRMPZdEkB180420.shtml

**Comment by higherdim** [comment]
Is it possible that Golaxy is a modified LZ trained with human pro games?

**Comment by pheasant75** [comment]
Sure, it is possible.  This is what open source is about.  People can
freely download and learn from it or modify it.  And we all learn from
alphaZero paper.
At least according to Golaxy team, they can change komi, handles handicap.
Definitely there is sth that LZ don't have.
